### ISSUE TITLE
Executing compaction in parallel between AVS directories

### ISSUE THREAD

**Issue by vstax** [leo-project/leofs#930]
Currently, when using multiple drives in JBOD configuration and having multiple `obj_containers.path` - one per drive, compaction still goes through all AVS files from first path, then through all AVS files from second path and so on. This makes the process much slower compared to going through all directories with AVS files at once (e.g. 4 times slower when using 4 drives) - or, alternatively, consume much more IO to go through at the same speed (lots of IO affecting just one drive slows down the whole system about as much as the same IO affecting each drive).

Instead, it should go through multiple AVS directories at once (thus running on multiple drives at once if there is AVS directory on each drive). This applies to both compaction and auto-compaction; I think that when having multiple more than one AVS directory, `compaction.limit_num_of_compaction_procs` and `autonomic_op.compaction.parallel_procs` should either be internally multiplied by amount of paths or just spread existing number of compactions over different paths the best way they can.

Example with multiplying: when having 4 AVS directories (A1, A2, A3, A4), compaction.limit_num_of_compaction_procs=4 and autonomic_op.compaction.is_enabled=2, compaction should process 16 AVS files in parallel (4 from each directory) and auto-compaction 8 AVS files in parallel (2 from each directory).  The problem is that old setting becomes a bit dishonest when having multiple AVS directories (but it can be just mentioned in config / documentation).

Alternatively (without multiplying), compaction could launch 4 AVS files in parallel (1 from each directory) and auto-compaction go through 2 AVS files from A1 + A2, then from A3 + A4, then again A1 + A2 and so on. The problem is that logic becomes more complicated, especially since AVS files from each drive might finish not at the same time, and it's too easy to get uneven load when specifying compaction.limit_num_of_compaction_procs=6 for example (which will start compacting 2 files at once from each A1 and A2 and 1 file at once from each A3 and A4). However, the current config file setting will be interpreted directly.

**Comment by mocchira** [comment]
@vstax sounds pretty rational to me. we will consider the spec in detail based on your comment.

**Comment by mocchira** [comment]
@vstax 

> Example with multiplying: when having 4 AVS directories (A1, A2, A3, A4), compaction.limit_num_of_compaction_procs=4 and autonomic_op.compaction.is_enabled=2, compaction should process 16 AVS files in parallel (4 from each directory) and auto-compaction 8 AVS files in parallel (2 from each directory). The problem is that old setting becomes a bit dishonest when having multiple AVS directories (but it can be just mentioned in config / documentation).

Sounds good to me rather than alternative one.
What do you think @yosukehara @windkit ?

**Comment by yosukehara** [comment]
@vstax @mocchira I agree with @vstax's idea too, and we need to write its documentation.

**Comment by mocchira** [comment]
#940 tried to solve the slow startup problem with parallel open between AVS directories however no improvement for some reason (still not sure at this time). so with the same approach, we might hit the same problem again so I'd like to postpone this until we will find out the root cause.

**Comment by vstax** [comment]
@mocchira I think the problem from #940 shouldn't affect this; I mean, even if there are some internal locks in (e)leveldb preventing parallel opening of containers, compaction already works in parallel just fine. Extra level of parallelism between AVS directories probably shouldn't change anything here?
